### PR TITLE
[Frontend][UST-135] Reputation History

### PR DESCRIPTION
**Pull request by DylanYang0523**
## Summary

- add date picker
- add reputation history

## Linked Issue

close https://github.com/social-tw/social-tw-website/issues/434

## Details

- ReputationHistory component
- SearchByDate component
- SearchDayLimitDialog component
- DatePicker component
- Table component
- EpochDateService
- useDatePicker hook

## Impacted Areas

- feature > profile
- feature > shared

## Tests

- EpochDateService.test

## Possible Impacts

- profile page

## Visual Materials

Screenshot 2024-07-22 at 00 05 24

## Verification Steps

No.

## Todo

Break down the progress of the PR for everyone to see what else you intend to include in this PR.

No.

## Checklist

-   [x] All new and existing tests pass
-   [x] I have updated the documentation (if applicable)
-   [x] My changes do not introduce new security vulnerabilities

### CODE REVIEW

File: packages/frontend/src/types/api.ts [hunk 1]
```diff
@@ -12,6 +12,15 @@ export enum SortKeys {
     VoteSum = 'voteSum',
 }
 
+export type FetchReputationHistoryResponse = {
+    score: string
```
Comment (Xiawpohr) [inline packages/frontend/src/types/api.ts:16]: The type of score should be number

File: packages/frontend/src/types/api.ts [hunk 2]
```diff
@@ -12,6 +12,15 @@ export enum SortKeys {
...
+    report: {
+        reason: string
```
Comment (Xiawpohr) [inline packages/frontend/src/types/api.ts:18]: `report.reason` means the reason content users write when they make a report. However, the reason column in UI shows why they get scores. You should use `type` in FetchReputationHistoryResponse.

Please check the types in relayer.

https://github.com/social-tw/social-tw-website/blob/12b196ba693037b2ea163fab8b5c3eddab48dbf5/packages/relay/src/types/Reputation.ts#L3-L19

File: packages/frontend/src/features/shared/hooks/useDatePicker.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        setStartDate(start)
+        setEndDate(end)
+        setIsDateSelected(!!start && !!end)
+    }
+    const setToday = () => {
+        const today = new Date()
+        setStartDate(today)
+        setEndDate(today)
+        setIsDateSelected(true)
+    }
```
Comment (Xiawpohr) [inline packages/frontend/src/features/shared/hooks/useDatePicker.ts:41]: There is an issue in this implementation. `fromEpoch` and `toEpoch` would be the same epoch number if `startDate` is equal to `endDate`. However, `fromEpcoh` should be the start epoch of today, and `toEpoch` should be the end epoch of today. `onChange`, `setPast7Days` and `setPast30Days` have the same issue.

`startDate` should be set to the start of today, and `endDate` should be the end of today. You can use day.js to implement this. For example,

```js
const now = dayjs()
setStartDate(now.startOf('day').toDate())
setEndDate(now.endOf('day').toDate())
```

Or you can fix this issue in `EpochDateService`.

File: packages/frontend/src/features/shared/hooks/useDatePicker.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    EpochDateService,
+    FromToEpoch,
+    InvalidFromToEpoch,
+} from '../services/EpochDateService'
+
+export function useDatePicker() {
+    const { userState } = useUserState()
+    const [startDate, setStartDate] = useState<undefined | Date>(undefined)
+    const [endDate, setEndDate] = useState<undefined | Date>(undefined)
+    const [isDateSelected, setIsDateSelected] = useState(false)
```
Comment (Xiawpohr) [inline packages/frontend/src/features/shared/hooks/useDatePicker.ts:13]: `isDateSelected` may be derived from `startDate` and `endDate`.

```js
const isDateSelected = !!startDate && !!endDate
```

File: packages/frontend/src/features/profile/hooks/useReputationHistory/useReputationHistory.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    FromToEpoch,
+    ValidFromToEpoch,
+} from '@/features/shared/services/EpochDateService'
+import { fetchReputationHistory } from '@/utils/api'
+import { useQuery } from '@tanstack/react-query'
+
+export function useReputationHistory(fromToEpoch: FromToEpoch) {
+    const { isFetching: isFetchingReputationHistory, data: reputationHistory } =
+        useQuery({
+            queryKey: [QueryKeys.HasSignedUp, fromToEpoch.from, fromToEpoch.to],
```
[the remark this answers is not in the document]
Comment (DylanYang0523) [reply packages/frontend/src/features/profile/hooks/useReputationHistory/useReputationHistory.ts:12]: Nice catch! Thank you 🙏🏼

**Review by Xiawpohr — APPROVED** [review]
Thanks

**Pull request merged**